### ISSUE TITLE
Dataset hangs and becomes nonresponsive when trying to load batch

### ISSUE THREAD

**Issue by jbohnslav** [NVIDIA/nvvl#89]
Hello, 

Thanks for all your amazing work on this project! It speeds up my training and inference by orders of magnitude. I'd love to incorporate it fully into my project, but the dataset hangs when trying to load a specific batch of a specific video. I think it might be related to [this issue](https://github.com/NVIDIA/nvvl/issues/29), but I'm not sure. 

Here's info about my system. I'm using the PyTorch 1.0 wrapper. 
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2018 NVIDIA Corporation
Built on Sat_Aug_25_21:08:01_CDT_2018
Cuda compilation tools, release 10.0, V10.0.130
```

When I use it in my training loop, due to shuffling it hangs at random times at random epochs. When I say hangs, I mean it becomes nonresponsive and often doesn't even respond to control-c. When I manage to control-c, the traceback says the KeyboardInterrupt at this line: 
```
File ".../nvvl/loader.py", line 111, in __next__
    self.dataset._finish_receive()
File ".../nvvl/dataset.py", line 347, in _finish_receive
   seq.wait_stream()
```

I managed to track it down to one position in one of my videos. I'll try re-encoding it with ffmpeg, but of course this is not ideal since it may occur in new data I get, and my users won't be able to debug it as easily. I set up a minimal example here:
```python
import nvvl
import os
from tqdm import tqdm_notebook as tqdm
basedir = r'/mnt/sdb/debug'
data_type = 'flow'
files = os.listdir(basedir)
files.sort()
flowfiles = [os.path.join(basedir, i) for i in files if 
            i.endswith(data_type+'.mp4')]
SEQUENCE_LENGTH = 10
dataset = nvvl.VideoDataset(flowfiles, SEQUENCE_LENGTH, device_id=0, log_level='debug')
print(len(dataset)) # 53943
# so that I can get debug information
_ = dataset[0]
for i in tqdm(range(50000, len(dataset))):
    string = '~~~ DATASET INDEX %d ~~~\n'%(i)
    ret = os.write(1, string.encode())
    batch = dataset[i]
```
Here's the output of the debugger on the first frame:
```
Opening file /mnt/sdb/debug/flow.mp4
Opened the first file, creating a video decoder
Using device: GeForce GTX 1080 Ti
Kernel module version 410.79, so using our own stream.
Got a request for /mnt/sdb/debug/flow.mp4 frame 0 send_queue_ has 0 frames left
Seeking to frame 0 timestamp 0
0: Sending   key  frame 0 to the decoder. size = 671 req.frame = 1 req.count = 9 nonkey_frame_count = 0
0: Sending nonkey frame 1 to the decoder. size = 661 req.frame = 1 req.count = 9 nonkey_frame_count = 1
Hardware Decoder Input Information
	Video codec     : AVC/H.264
	Frame rate      : 2997/100 = 29.97 fps
	Sequence format : Progressive
	Coded frame size: [272, 240]
	Display area    : [0, 0, 270, 240]
	Chroma format   : YUV 420
	Bit depth       : 8
NVDEC Capabilities
	Max width : 4096
	Max height : 4096
	Max MB count : 65536
	Min width : 48
	Min height :16
	Using full size : [0, 0], [270, 240]
```

And finally, here's where the error occurs. It's always at dataset index 53942, right near the end of the video. 
```
~~~ DATASET INDEX 53942 ~~~
Got a request for /mnt/sdb/debug/flow.mp4 frame 53942 send_queue_ has 0 frames left
Seeking to frame 53942 timestamp 21576800
0: We got ahead of ourselves! 53943 > 53942 + 0 seek_hack = 1
Seeking to frame 53940 timestamp 21576000
popping frame (0/10) 0 reqs left
0: Sending   key  frame 53936 to the decoder. size = 1361 req.frame = 53942 req.count = 10 nonkey_frame_count = 0
0: Sending nonkey frame 53937 to the decoder. size = 900 req.frame = 53942 req.count = 10 nonkey_frame_count = 0
Sending a picture for decode size: 1361 pic index: 4
0: Sending nonkey frame 53938 to the decoder. size = 945 req.frame = 53942 req.count = 10 nonkey_frame_count = 0
Sending a picture for decode size: 900 pic index: 18
0: Sending nonkey frame 53939 to the decoder. size = 799 req.frame = 53942 req.count = 10 nonkey_frame_count = 0
Sending a picture for decode size: 945 pic index: 3
Ditching frame 53936 since we are waiting for frame 53942
0: Sending nonkey frame 53940 to the decoder. size = 865 req.frame = 53942 req.count = 10 nonkey_frame_count = 0
Sending a picture for decode size: 799 pic index: 9
Ditching frame 53937 since we are waiting for frame 53942
0: Sending nonkey frame 53941 to the decoder. size = 728 req.frame = 53942 req.count = 10 nonkey_frame_count = 0
Sending a picture for decode size: 865 pic index: 2
Ditching frame 53938 since we are waiting for frame 53942
0: Sending nonkey frame 53942 to the decoder. size = 812 req.frame = 53942 req.count = 10 nonkey_frame_count = 1
Sending a picture for decode size: 728 pic index: 7
Ditching frame 53939 since we are waiting for frame 53942
0: Sending   key  frame 53943 to the decoder. size = 862 req.frame = 53944 req.count = 8 nonkey_frame_count = 0
Sending a picture for decode size: 812 pic index: 10
Ditching frame 53940 since we are waiting for frame 53942
0: Sending nonkey frame 53944 to the decoder. size = 967 req.frame = 53944 req.count = 8 nonkey_frame_count = 1
Ditching frame 53941 since we are waiting for frame 53942
Going ahead with frame 53942 wanted count: 10 disp_info->picture_index: 10
Sending a picture for decode size: 862 pic index: 14
0: Sending nonkey frame 53945 to the decoder. size = 1162 req.frame = 53944 req.count = 8popping frame ( nonkey_frame_count = 21/
10) 0 reqs left
Sending a picture for decode size: 967 pic index: 16
0: Sending nonkey frame 53946 to the decoder. size = 1095 req.frame = 53944 req.count = 8 nonkey_frame_count = 3
Sending a picture for decode size: 1162 pic index: 6
Going ahead with frame 53943 wanted count: 9 disp_info->picture_index: 14
0: Sending nonkey frame 53947 to the decoder. size = 1207 req.frame = 53944 req.count = 8 nonkey_frame_count = 4
Sending a picture for decode size: 1095 pic index: 8
Going ahead with frame 53944 wanted count: 8 disp_info->picture_index: 16
0: Sending nonkey frame 53948 to the decoder. size = 1216 req.frame = 53944 req.count = 8 nonkey_frame_count = 5
Sending a picture for decode size: 1207 pic index: 12
popping frame (2/10) 1 reqs left
Going ahead with frame 53945 wanted count: 7 disp_info->picture_index: 6
0: Sending nonkey frame 53949 to the decoder. size = 1240 req.frame = 53944 req.count = 8 nonkey_frame_count = 6
Sending a picture for decode size: 1216 pic index: 11
Going ahead with frame 53946 wanted count: 6 disp_info->picture_index: 8
popping frame (03: Sending /nonkey frame 1053950 to the decoder. size = 264 req.frame = 53944)  req.count = 28 reqs left nonkey_frame_count = 
7
Sending a picture for decode size: 1240 pic index: 0
Going ahead with frame 53947 wanted count: 5 disp_info->picture_index: 12
Sending a picture for decode size: 263 pic index: 19
Going ahead with frame 53948 wanted count: 4 disp_info->picture_index: 11
Going ahead with frame 53949 wanted count: 3 disp_info->picture_index: 0
Going ahead with frame 53950 wanted count: 2 disp_info->picture_index: 19
popping frame (4/10) 5 reqs left
popping frame (5/10) 4 reqs left
popping frame (6/10) 3 reqs left
popping frame (7/10) 2 reqs left
popping frame (8/10) 1 reqs left
popping frame (9/10) 0 reqs left
```

As far as I can tell, there's a mismatch between the number of frames to pop and the number of reqs.
When I use sequence_length=1, it still hangs, but the debug is different:
```
~~~ DATASET INDEX 53951 ~~~
Got a request for /mnt/sdb/debug/flow.mp4 frame 53951 send_queue_ has 0 frames left
Seeking to frame 53951 timestamp 21580400
0: Sending   key  frame 53943 to the decoder. size = 862 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
0: Sending nonkey frame 53944 to the decoder. size = 967 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 862 pic index: 19
0: Sending nonkey frame 53945 to the decoder. size = 1162 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 967 pic index: 8
0: Sending nonkey frame 53946 to the decoder. size = 1095 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 1162 pic index: 3
Ditching frame 53943 since we are waiting for frame 53951
0: Sending nonkey frame 53947 to the decoder. size = 1207 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 1095 pic index: 2
popping frame (0/1) Ditching frame 0 reqs left53944 since we are waiting for frame 
53951
0: Sending nonkey frame 53948 to the decoder. size = 1216 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 1207 pic index: 4
Ditching frame 53945 since we are waiting for frame 53951
0: Sending nonkey frame 53949 to the decoder. size = 1240 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 1216 pic index: 7
Ditching frame 53946 since we are waiting for frame 53951
0: Sending nonkey frame 53950 to the decoder. size = 264 req.frame = 53951 req.count = 1 nonkey_frame_count = 0
Sending a picture for decode size: 1240 pic index: 9
Ditching frame 53947 since we are waiting for frame 53951
Sending a picture for decode size: 263 pic index: 13
Ditching frame 53948 since we are waiting for frame 53951
Ditching frame 53949 since we are waiting for frame 53951
Ditching frame 53950 since we are waiting for frame 53951
```

Thanks for all your help, and hopefully we can track down this issue!

**Comment by akssri** [comment]
This might be an issue with ffmpeg itself. Doing inaccurate seek with ffmpeg (using time instead of frame) towards the end of the video has the same effect.

**Comment by bomri** [comment]
I am having a similar issue, running the iterator in a for-loop it keeps hanging, when I'm adding a short sleep (time.sleep(0.005)) between iterations it seems to solve the problem.
could be a syncing problem with the c++ code under the hood.
